***Exam 2***


Exam 2 Author: Kayce Johnson

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

from itables import show

# This stops a few warning messages from showing
pd.options.mode.chained_assignment = None 
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [4]:
# Load the data.
file_name = 'flights'
file_location = 'https://joannabieri.com/introdatascience/data/'+file_name+'.csv'
DF = pd.read_csv(file_location)
DF


,year,month,day,dep_time,sched_dep_time,dep_delay,arr_time,sched_arr_time,arr_delay,carrier,flight,tailnum,origin,dest,air_time,distance,hour,minute,time_hour
0,2013,1,1,517.0,515,2.0,830.0,819,11.0,UA,1545,N14228,EWR,IAH,227.0,1400.0,5.0,15.0,2013-01-01 10:00:00
1,2013,1,1,533.0,529,4.0,850.0,830,20.0,UA,1714,N24211,LGA,IAH,227.0,1416.0,5.0,29.0,2013-01-01 10:00:00
2,2013,1,1,542.0,540,2.0,923.0,850,33.0,AA,1141,N619AA,JFK,MIA,160.0,1089.0,5.0,40.0,2013-01-01 10:00:00
3,2013,1,1,544.0,545,-1.0,1004.0,1022,-18.0,B6,725,N804JB,JFK,BQN,183.0,1576.0,5.0,45.0,2013-01-01 10:00:00
4,2013,1,1,554.0,600,-6.0,812.0,837,-25.0,DL,461,N668DN,LGA,ATL,116.0,762.0,6.0,0.0,2013-01-01 11:00:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
336771,2013,9,30,NaN,1455,NaN,NaN,1634,NaN,9E,3393,NaN,JFK,DCA,NaN,213.0,14.0,55.0,2013-09-30 18:00:00
336772,2013,9,30,NaN,2200,NaN,NaN,2312,NaN,9E,3525,NaN,LGA,SYR,NaN,198.0,22.0,0.0,2013-10-01 02:00:00
336773,2013,9,30,NaN,1210,NaN,NaN,1330,NaN,MQ,3461,N535MQ,LGA,BNA,NaN,764.0,12.0,10.0,2013-09-30 16:00:00
336774,2013,9,30,NaN,1159,NaN,NaN,1344,NaN,MQ,3572,N511MQ,LGA,CLE,NaN,419.0,11.0,59.0,2013-09-30 15:00:00


In [6]:
file_name = 'airports'
file_location = 'https://joannabieri.com/introdatascience/data/'+file_name+'.csv'
DF_airports = pd.read_csv(file_location)
show(DF_airports)


Loading ITables v2.6.2 from the internet... (need help?)


In [7]:
DF.describe()

,year,month,day,dep_time,sched_dep_time,dep_delay,arr_time,sched_arr_time,arr_delay,flight,air_time,distance,hour,minute
count,336776.0,336776.000000,336776.000000,328521.000000,336776.000000,328521.000000,328063.000000,336776.000000,327346.000000,336776.000000,327346.000000,336776.000000,336776.000000,336776.000000
mean,2013.0,6.548510,15.710787,1349.109947,1344.254840,12.639070,1502.054999,1536.380220,6.895377,1971.923620,150.686460,1039.912604,13.180247,26.230100
std,0.0,3.414457,8.768607,488.281791,467.335756,40.210061,533.264132,497.457142,44.633292,1632.471938,93.688305,733.233033,4.661316,19.300846
min,2013.0,1.000000,1.000000,1.000000,106.000000,-43.000000,1.000000,1.000000,-86.000000,1.000000,20.000000,17.000000,1.000000,0.000000
25%,2013.0,4.000000,8.000000,907.000000,906.000000,-5.000000,1104.000000,1124.000000,-17.000000,553.000000,82.000000,502.000000,9.000000,8.000000
50%,2013.0,7.000000,16.000000,1401.000000,1359.000000,-2.000000,1535.000000,1556.000000,-5.000000,1496.000000,129.000000,872.000000,13.000000,29.000000
75%,2013.0,10.000000,23.000000,1744.000000,1729.000000,11.000000,1940.000000,1945.000000,14.000000,3465.000000,192.000000,1389.000000,17.000000,44.000000
max,2013.0,12.000000,31.000000,2400.000000,2359.000000,1301.000000,2400.000000,2359.000000,1272.000000,8500.000000,695.000000,4983.000000,23.000000,59.000000


There are 336776 observations across 19 variables. There are a lot on NaNs - particularly in air time, deploy time, and deployed delay - which feels like flights that never left. 

Most variables are numerical, except the ailnum,	origin, and	dest. The time variable will need some work before it can be treated as a numerical variable.

***Question 1***

In [8]:
my_columns = ['origin', 'dest']
DF1 = DF[my_columns]
DF1

show(DF1.groupby('origin')['dest'].value_counts())
show(DF1.value_counts())

Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


Page 1 shows the top 10 destinations for EWR. JFK starts on page 9, and LGA on page 16. 

Top 10 for EWR: ORD, BOS, SFO, CLT, STL, MCO, LAX, IAH, FLL, DTW

Top 10 for JFK: LAX, SFO, BOS, MCO, SJU, FLL, LAS, BUF, MIA, DCA

Top 10 for LGA: ATL, ORD, CLT, MIA, DTW, DFW, DCA, BOS, FLL, MSP

For each, it looks like their top destinations are in different parts of the county, but with a lot of major overlap to major cities throughout the United States, primarily.

The top 10 TOTAL: LAX, ATL, ORD, SFO, CLT, ORD, BOS, MIA, MCO, BOS

Its very interseting that ORD is listed twice - once from EWR and once from LGA. That means its the highest total destination, across two airports.

***Question 2***

In [9]:

my_new_columns = ['origin', 'carrier']
DF2 = DF[my_new_columns]
DF2

show(DF2.groupby('origin')['carrier'].value_counts())
show(DF2.value_counts(ascending= False))

Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


UA (United Airlines)
EV (ExpressJet Airlines)
B6 (JetBlue Airways)
DL (Delta Air Lines Inc.)
MQ (Envoy Air)
AA (American Airlines)
9E (Endeavor Air Inc.)
US (US Airways Inc.)

These airlines have the highest number of departures of flights from New York airports in 2013. (Less than 10 total airlines due to doubling up of some airport destinations from different origins).

In [10]:
file_name = 'airlines'
file_location = 'https://joannabieri.com/introdatascience/data/'+file_name+'.csv'
DF_airlines = pd.read_csv(file_location)
show(DF_airlines)

Loading ITables v2.6.2 from the internet... (need help?)


***Question 3***

In [11]:
DF['arr_delay'].replace('NaN','NA',inplace=True)
filter_columns = ['arr_delay', 'carrier']
DF = DF[DF['arr_delay'].notna()]
DF_filter = DF[filter_columns]


DF_filter.groupby('carrier').mean()

,arr_delay
carrier,
9E,7.379669
AA,0.364291
AS,-9.930889
B6,9.457973
DL,1.644341
EV,15.796431
F9,21.920705
FL,20.115906
HA,-6.915205


a. Highest mean arrival delay - F9 (Frontier Airlines) with an average of 21.9 minutes.

b. Lowest mean arrival delay - AS (Alaska Airlines) with an average -9.9 minute delay meaning they, on average, arrive 9 minutes earlier than scheduled. 

***Question 4***


In [24]:
DF['dep_time'].replace('NaN','NA',inplace=True)
DF = DF[DF['dep_time'].notna()]

mask_0_6 = (DF['dep_time'] >= 0) & (DF['dep_time'] <= 600)
mask_6_12 = (DF['dep_time'] > 600) & (DF['dep_time'] <= 1200)
mask_12_18 = (DF['dep_time'] > 1200) & (DF['dep_time'] <= 1800)
mask_18_24 = (DF['dep_time'] > 1800) & (DF['dep_time'] <= 2359)

DF_0_6 = DF[mask_0_6]
DF_6_12 = DF[mask_6_12]
DF_12_18 = DF[mask_12_18]
DF_18_24 = DF[mask_18_24]

dep_columns = ['dep_time', 'dep_delay']



#above - actual departure time
#below - scheduled departure time

DF['sched_dep_time'].replace('NaN','NA',inplace=True)
DF = DF[DF['sched_dep_time'].notna()]

mask_0_6s = (DF['sched_dep_time'] >= 0) & (DF['sched_dep_time'] <= 600)
mask_6_12s = (DF['sched_dep_time'] > 600) & (DF['sched_dep_time'] <= 1200)
mask_12_18s = (DF['sched_dep_time'] > 1200) & (DF['sched_dep_time'] <= 1800)
mask_18_24s = (DF['sched_dep_time'] > 1800) & (DF['sched_dep_time'] <= 2359)

DF_0_6s = DF[mask_0_6]
DF_6_12s = DF[mask_6_12]
DF_12_18s = DF[mask_12_18]
DF_18_24s = DF[mask_18_24]

sched_dep_columns = ['sched_dep_time', 'dep_delay']

show(DF_0_6[dep_columns].mean())
show(DF_6_12[dep_columns].mean())
show(DF_12_18[dep_columns].mean())
show(DF_18_24[dep_columns].mean())

show(DF_0_6s[dep_columns].mean())
show(DF_6_12s[dep_columns].mean())
show(DF_12_18s[dep_columns].mean())
show(DF_18_24s[dep_columns].mean())


Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


Loading ITables v2.6.2 from the internet... (need help?)


The highest average departure delay by actual departure time is during the last 6 hours of the day.




***Question 5***

In [14]:
DF.sort_values('air_time', ascending = False)


,year,month,day,dep_time,sched_dep_time,dep_delay,arr_time,sched_arr_time,arr_delay,carrier,flight,tailnum,origin,dest,air_time,distance,hour,minute,time_hour
151467,2013,3,17,1337.0,1335,2.0,1937.0,1836,61.0,UA,15,N77066,EWR,HNL,695.0,4963.0,13.0,35.0,2013-03-17 17:00:00
115752,2013,2,6,853.0,900,-7.0,1542.0,1540,2.0,HA,51,N389HA,JFK,HNL,691.0,4983.0,9.0,0.0,2013-02-06 14:00:00
149618,2013,3,15,1001.0,1000,1.0,1551.0,1530,21.0,HA,51,N388HA,JFK,HNL,686.0,4983.0,10.0,0.0,2013-03-15 14:00:00
151311,2013,3,17,1006.0,1000,6.0,1607.0,1530,37.0,HA,51,N380HA,JFK,HNL,686.0,4983.0,10.0,0.0,2013-03-17 14:00:00
150547,2013,3,16,1001.0,1000,1.0,1544.0,1530,14.0,HA,51,N384HA,JFK,HNL,683.0,4983.0,10.0,0.0,2013-03-16 14:00:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
336771,2013,9,30,NaN,1455,NaN,NaN,1634,NaN,9E,3393,NaN,JFK,DCA,NaN,213.0,14.0,55.0,2013-09-30 18:00:00
336772,2013,9,30,NaN,2200,NaN,NaN,2312,NaN,9E,3525,NaN,LGA,SYR,NaN,198.0,22.0,0.0,2013-10-01 02:00:00
336773,2013,9,30,NaN,1210,NaN,NaN,1330,NaN,MQ,3461,N535MQ,LGA,BNA,NaN,764.0,12.0,10.0,2013-09-30 16:00:00
336774,2013,9,30,NaN,1159,NaN,NaN,1344,NaN,MQ,3572,N511MQ,LGA,CLE,NaN,419.0,11.0,59.0,2013-09-30 15:00:00


Index - 151467

In [22]:
file_name = 'planes'
file_location = 'https://joannabieri.com/introdatascience/data/'+file_name+'.csv'
DF_planes = pd.read_csv(file_location)
DF_planes

row = DF_planes[DF_planes['tailnum'] == 'N77066']
show(row)


Loading ITables v2.6.2 from the internet... (need help?)


a. This flight was 1937 minutes long, or 32.38 hours (really??)

b. That flight flew to HNL ( )
**I am sure this is Honolulu- why is it not in the airports data set?**

c. This flight had 292 seats.


***Question 6***

In [23]:
DF_airports

,faa,name,lat,lon,alt,tz,dst,tzone
0,04G,Lansdowne Airport,41.130472,-80.619583,1044.0,-5.0,A,America/New_York
1,06A,Moton Field Municipal Airport,32.460572,-85.680028,264.0,-6.0,A,America/Chicago
2,06C,Schaumburg Regional,41.989341,-88.101243,801.0,-6.0,A,America/Chicago
3,06N,Randall Airport,41.431912,-74.391561,523.0,-5.0,A,America/New_York
4,09J,Jekyll Island Airport,31.074472,-81.427778,11.0,-5.0,A,America/New_York
...,...,...,...,...,...,...,...,...
1453,ZUN,Black Rock,35.083228,-108.791778,6454.0,-7.0,A,America/Denver
1454,ZVE,New Haven Rail Station,41.298669,-72.925992,7.0,-5.0,A,America/New_York
1455,ZWI,Wilmington Amtrak Station,39.736667,-75.551667,0.0,-5.0,A,America/New_York
1456,ZWU,Washington Union Station,38.897460,-77.006430,76.0,-5.0,A,America/New_York


**Does the contiguous US long/ lat remove the parts of Canada and South America? How do we combat this?

***Question 7***

In [25]:
DF['arrival'] = DF['arr_delay'].apply(lambda x: 'ontime' if x <= 0 else 'delayed')
DF
#using the filtered data with the removed the NA values
#This bit I found help with on Stack Overflow


,year,month,day,dep_time,sched_dep_time,dep_delay,arr_time,sched_arr_time,arr_delay,carrier,flight,tailnum,origin,dest,air_time,distance,hour,minute,time_hour,arrival
0,2013,1,1,517.0,515,2.0,830.0,819,11.0,UA,1545,N14228,EWR,IAH,227.0,1400.0,5.0,15.0,2013-01-01 10:00:00,delayed
1,2013,1,1,533.0,529,4.0,850.0,830,20.0,UA,1714,N24211,LGA,IAH,227.0,1416.0,5.0,29.0,2013-01-01 10:00:00,delayed
2,2013,1,1,542.0,540,2.0,923.0,850,33.0,AA,1141,N619AA,JFK,MIA,160.0,1089.0,5.0,40.0,2013-01-01 10:00:00,delayed
3,2013,1,1,544.0,545,-1.0,1004.0,1022,-18.0,B6,725,N804JB,JFK,BQN,183.0,1576.0,5.0,45.0,2013-01-01 10:00:00,ontime
4,2013,1,1,554.0,600,-6.0,812.0,837,-25.0,DL,461,N668DN,LGA,ATL,116.0,762.0,6.0,0.0,2013-01-01 11:00:00,ontime
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
336765,2013,9,30,2240.0,2245,-5.0,2334.0,2351,-17.0,B6,1816,N354JB,JFK,SYR,41.0,209.0,22.0,45.0,2013-10-01 02:00:00,ontime
336766,2013,9,30,2240.0,2250,-10.0,2347.0,7,-20.0,B6,2002,N281JB,JFK,BUF,52.0,301.0,22.0,50.0,2013-10-01 02:00:00,ontime
336767,2013,9,30,2241.0,2246,-5.0,2345.0,1,-16.0,B6,486,N346JB,JFK,ROC,47.0,264.0,22.0,46.0,2013-10-01 02:00:00,ontime
336768,2013,9,30,2307.0,2255,12.0,2359.0,2358,1.0,B6,718,N565JB,JFK,BOS,33.0,187.0,22.0,55.0,2013-10-01 02:00:00,delayed


In [32]:
BOXPLOT??

Object `BOXPLOT` not found.
